contracts/src/instances/morpho-blue/MorphoBlueBase.sol

```diff
@@ -0,0 +1,221 @@
+import { IHyperdrive } from "../../interfaces/IHyperdrive.sol";
+import { IMorphoBlueHyperdrive } from "../../interfaces/IMorphoBlueHyperdrive.sol";
+import { HyperdriveBase } from "../../internal/HyperdriveBase.sol";
+import { MorphoBlueConversions } from "./MorphoBlueConversions.sol";
+
+/// @author DELV
+/// @title MorphoBlueBase
+/// @notice The base contract for the MorphoBlue Hyperdrive implementation.
+/// @dev This Hyperdrive implementation is designed to work with standard
+///      MorphoBlue vaults. Non-standard implementations may not work correctly
+///      and should be carefully checked.
+/// @custom:disclaimer The language used in this code is for coding convenience
+///                    only, and is not intended to, and does not, have any
+///                    particular legal or regulatory significance.
+abstract contract MorphoBlueBase is HyperdriveBase {
+    using SafeERC20 for ERC20;
+    using MarketParamsLib for MarketParams;
+    using MorphoBalancesLib for IMorpho;
+    using SharesMathLib for uint256;
+
+    /// @dev The Morpho Blue contract.
+    IMorpho internal immutable _vault;
+
+    /// @dev The collateral token for this Morpho Blue market.
+    address internal immutable _collateralToken;
+
+    /// @dev The oracle for this Morpho Blue market.
+    address internal immutable _oracle;
+
+    /// @dev The IRM for this Morpho Blue market.
+    address internal immutable _irm;
+
+    /// @dev The LLTV for this Morpho Blue market.
+    uint256 internal immutable _lltv;
+
+    /// @notice Instantiates the MorphoBlueHyperdrive base contract.
+    /// @param _params The Morpho Blue params.
+    constructor(IMorphoBlueHyperdrive.MorphoBlueParams memory _params) {
+        // Initialize the Morpho vault immutable.
+        _vault = _params.morpho;
+
+        // Initialize the market parameters immutables. We don't need an
+        // immutable for the loan token because we set the base token to the
+        // loan token.
+        _collateralToken = _params.collateralToken;
+        _oracle = _params.oracle;
+        _irm = _params.irm;
+        _lltv = _params.lltv;
+
+        // Approve the Morpho vault with 1 wei. This ensures that all of the
+        // subsequent approvals will be writing to a dirty storage slot.
+        ERC20(address(_baseToken)).forceApprove(address(_vault), 1);
+    }
+
+    /// Yield Source ///
+
+    /// @dev Accepts a deposit from the user in base.
+    /// @param _baseAmount The base amount to deposit.
+    /// @param _extraData Additional data to pass to the Morpho vault. This
+    ///        should be zero if it is unused.
+    /// @return The shares that were minted in the deposit.
+    /// @return The amount of ETH to refund. Since this yield source isn't
+    ///         payable, this is always zero.
+    function _depositWithBase(
+        uint256 _baseAmount,
+        bytes calldata _extraData
+    ) internal override returns (uint256, uint256) {
```

Comment:
nit: I feel like any function with multiple return values should use named values (especially if they're the same type)

Reply:
I like named returns, but I don't agree with this in general. For functions like `openShort`, we need the flexibility to manage the stack more carefully than this would allow. In this case, I'm fine with using named returns, so I'll add them.